syntax = "proto3";

package GenericInteractSpace;
option java_multiple_files = true;
option java_package = "org.eclipse.mosaic.fed.output.generator.zeromq";
option java_outer_classname = "GenericInteractProtos";
option optimize_for = LITE_RUNTIME;


message GenericInteract {
   int64 time = 1;
   int64 id = 2;
   string sender_id = 3;
   string type_id = 4;

   message RsuRegistrationBuf {
      string name = 1;
      repeated double position = 2;
   }

   message VehicleDataBuf {
      int64 time = 1;
      string name = 2;
      repeated double position = 3;
      // IRoadPosition start
      //IRoadPosition IConnnection start
      string road_id = 4;
      // IRoadPosition IConnnection end
      // IRoadPosition end
      string route_id = 5;
      double speed = 6;
      double longitudinal_acc = 7;
   }


   message VehicleUpdatesBuf {
      int64 time = 1;
      repeated VehicleDataBuf added = 2;
      repeated VehicleDataBuf updated = 3;
   }


   message VehicleRegistrationBuf {
      string name = 1;
      string group = 2;
      repeated string applications = 3;
   }

   message V2xMessageReceptionBuf {
      string receiver_name = 1;
      int64 message_id = 2;
      int64 send_time = 3;
      int64 receive_time = 4;
      double receive_signal_strength = 5;
   }

   oneof Interaction {
      RsuRegistrationBuf rsu_registration = 5;
      VehicleUpdatesBuf vehicle_updates = 6;
      VehicleRegistrationBuf vehicle_registration = 7;
      V2xMessageReceptionBuf v2x_message_reception = 8;
   }
}
